io/src/lib.rs

```diff
@@ -230,6 +230,59 @@ impl<T: AsRef<[u8]>> BufRead for Cursor<T> {
     }
 }
 
+/// A [`Read`]er which keeps an internal buffer to avoid hitting the underlying stream directly for
+/// every read, implementing [`BufRead`].
+pub struct BufReader<'a, R: Read> {
+    inner: &'a mut R,
+    buf: [u8; 4096],
+    pos: usize,
+    limit: usize,
+}
+
+impl<'a, R: Read> BufReader<'a, R> {
+    /// Creates a [`BufReader`] which will read from the given `inner`.
+    pub fn new(inner: &'a mut R) -> Self {
+        BufReader {
+            inner,
+            buf: [0; 4096],
+            pos: 0,
+            limit: 0,
+        }
+    }
+}
+
+impl<'a, R: Read> Read for BufReader<'a, R> {
+    #[inline]
+    fn read(&mut self, output: &mut [u8]) -> Result<usize> {
+        let input = self.fill_buf()?;
+        let count = cmp::min(input.len(), output.len());
+        output[..count].copy_from_slice(&input[..count]);
+        self.consume(count);
+        Ok(count)
+    }
+}
+
+impl<'a, R: Read> BufRead for BufReader<'a, R> {
+    #[inline]
+    fn fill_buf(&mut self) -> Result<&[u8]> {
+        if self.pos < self.limit {
+            Ok(&self.buf[self.pos..self.limit])
+        } else {
+            let count = self.inner.read(&mut self.buf[..])?;
```

Comment:
Nit: the indexing shouldn't be needed since `&mut [T; N]` coerces to `&mut [T]`.